lib/connectors/memory.js

```diff
@@ -250,6 +250,46 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
+    console.log (element);
+    var filter = { where: where };
+    console.log(filter);
+    console.log(model);
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+    var primaryKey = this.idName(model);
+    self.all(model, filter, options, function(err, instances) {
+      if (err) callback(err);
+      else {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(primaryKey)) {
+          self.create(model, data, options, function(err, id) {
+            self.setIdValue(model, data, id);
+            callback(err, data, { isNewInstance: true });
+          });
+        } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
```

Comment:
As mentioned elsewhere `id` can be auto generated id in many cases and not provided in `data` object; you should take care of this scenario.